### PR TITLE
vim-patch:8.2.3385,8.2.3393

### PR DESCRIPTION
**Pull request by jasonccox**
#### vim-patch:8.2.3385: escaping for fish shell does not work properly

Problem:    Escaping for fish shell does not work properly.
Solution:   Insert a backslash before a backslash. (Jason Cox, closes vim/vim#8810)
https://github.com/vim/vim/commit/6e82351130ddb8d13cf3748b47f07cae77886fc7

#### vim-patch:8.2.3393: escaping for fish shell is skipping some characters

Problem:    Escaping for fish shell is skipping some characters.
Solution:   Escape character after backslash if needed. (Jason Cox,
            closes vim/vim#8827)
https://github.com/vim/vim/commit/6631597452d4644f485a09e4036d117e5f91de70

### CODE REVIEW

File: src/nvim/option.c [hunk 1]
```diff
@@ -7618,6 +7618,12 @@ int csh_like_shell(void)
   return strstr((char *)path_tail(p_sh), "csh") != NULL;
 }
 
+/// Return true when 'shell' has "fish" in the tail.
+int fish_like_shell(void)
```
Comment (seandewar) [inline src/nvim/option.c:7622]:
```suggestion
bool fish_like_shell(void)
```

File: src/nvim/strings.c [hunk 2]
```diff
@@ -190,13 +190,18 @@ char_u *vim_strsave_shellescape(const char_u *string,
   char_u      *escaped_string;
   size_t l;
   int csh_like;
+  int fish_like;
```
Comment (seandewar) [inline src/nvim/strings.c:193]:
```suggestion
  bool fish_like;
```

File: src/nvim/strings.c [hunk 3]
```diff
@@ -190,13 +190,18 @@ char_u *vim_strsave_shellescape(const char_u *string,
...
 
   /* Only csh and similar shells expand '!' within single quotes.  For sh and
    * the like we must not put a backslash before it, it will be taken
    * literally.  If do_special is set the '!' will be escaped twice.
    * Csh also needs to have "\n" escaped twice when do_special is set. */
   csh_like = csh_like_shell();
 
+  /* Fish shell uses '\' as an escape character within single quotes, so '\'
+   * itself must be escaped to get a literal '\'. */
```
Comment (seandewar) [inline src/nvim/strings.c:202]:
```suggestion
  // Fish shell uses '\' as an escape character within single quotes, so '\'
  // itself must be escaped to get a literal '\'.
```

File: src/nvim/strings.c [hunk 4]
```diff
@@ -220,6 +225,9 @@ char_u *vim_strsave_shellescape(const char_u *string,
       ++length;                         /* insert backslash */
       p += l - 1;
     }
+    if (*p == '\\' && fish_like) {
+      ++length;                         /* insert backslash */
```
Comment (seandewar) [inline src/nvim/strings.c:229]:
```suggestion
      length++;  // insert backslash
```

**Review by seandewar — CHANGES_REQUESTED** [review]
Just some style things `make lint` should've caught.
Also, for strictly boolean values, prefer `bool`. :+1:
(Don't worry about the surrounding code; it hasn't been updated yet)

I can't speak for the tests, though.

**Comment by jasonccox** [comment]
Ah, didn't realize `make lint` existed. Those should all be fixed now.

File: src/nvim/strings.c [hunk 5]
```diff
@@ -220,6 +225,9 @@ char_u *vim_strsave_shellescape(const char_u *string,
       ++length;                         /* insert backslash */
       p += l - 1;
     }
+    if (*p == '\\' && fish_like) {
+      length++; // insert backslash
```
Comment (seandewar) [inline src/nvim/strings.c:229]:
```suggestion
      length++;  // insert backslash
```

(2 spaces before a comment)

Comment (jasonccox) [reply src/nvim/strings.c]: Whoops, I'm on a different computer and couldn't get `make lint` actually running, so I missed this one. Fixed now.

**Comment by jasonccox** [comment]
Is there anything else I need to change here before getting this merged in?

**Pull request merged**